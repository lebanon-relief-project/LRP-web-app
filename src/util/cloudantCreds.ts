import { CloudantCredentials } from "../types/Cloudant";

export function getCredentials(): CloudantCredentials {
  const VCAP = JSON.parse(process.env.VCAP_SERVICES);

  if (!VCAP || !VCAP.cloudantNoSQLDB) {
    throw new Error("CLOUDANT_CREDENTIALS missing");
  }

  const apiKey = VCAP.cloudantNoSQLDB[0].credentials.apikey;
  const host = VCAP.cloudantNoSQLDB[0].credentials.url;

  return {
    url: `${host}`,
    apiKey: `${apiKey}`,
  };
}
