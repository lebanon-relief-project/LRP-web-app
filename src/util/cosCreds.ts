import { COS_URL } from "../statics";
import { CosCredentials } from "../types/Cos";

export function getCosCredentials(): CosCredentials {

  const VCAP = JSON.parse(process.env.VCAP_SERVICES);

  if (!VCAP || !VCAP["cloud-object-storage"]) {
    throw new Error("COS credentials missing");
  }
  
  const endpoint = COS_URL;
  const apikey = VCAP["cloud-object-storage"][0].credentials.apikey;
  const resourceInstanceId =
    VCAP["cloud-object-storage"][0].credentials["resource_instance_id"];
  const accessKeyId =
    VCAP["cloud-object-storage"][0].credentials["cos_hmac_keys"][
      "access_key_id"
    ];
  const secretAccessKey =
    VCAP["cloud-object-storage"][0].credentials["cos_hmac_keys"][
      "secret_access_key"
    ];

  return {
    endpoint: endpoint,
    apiKeyId: apikey,
    serviceInstanceId: resourceInstanceId,
    signatureVersion: "v4",
    accessKeyId: accessKeyId,
    secretAccessKey: secretAccessKey,
  };
}
