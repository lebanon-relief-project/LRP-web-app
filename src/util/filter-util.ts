import { FilterType } from "src/types/Filter";

export function sanitizeFilter(unSafeFilter: any): FilterType {
  const schema: Record<keyof FilterType, string> = {
    languages: "string",
  };

  let cleanFilter: FilterType = {};

  Object.keys(schema).forEach((key) => {
    if (unSafeFilter[key]) {
      cleanFilter[key] = unSafeFilter[key];

      if (typeof schema[key] === "string") {
        if (typeof unSafeFilter[key] === "string")
          cleanFilter[key] = [unSafeFilter[key]];
        cleanFilter[key] = cleanFilter[key].map((entry) => {
          return entry.toLowerCase();
        });
      }

      if (typeof schema[key] === "object") {
        cleanFilter[key] = JSON.parse(cleanFilter[key]);
      }
    }
  });

  return cleanFilter;
}
