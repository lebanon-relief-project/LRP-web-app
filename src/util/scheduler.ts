import Container from "typedi";
import { CloudantService, FlashCardService } from "../services";
import { CronJob } from "cron";

var flashCardService: FlashCardService = Container.get("FlashCardService");

export function scheduleJob(): void {
  var job = new CronJob(
    "* 22 2 * * *",
    async function () {
      console.log("counting started");
      try {
        let result = await flashCardService.countSelections();
        if (result) {
          console.log("counting executed successfully");
        }
      } catch (error) {
        console.error(error);
      }
    },
    null,
    false,
    "Europe/London"
  );
  job.start();
  console.log("job was scheduled successfully");
}
