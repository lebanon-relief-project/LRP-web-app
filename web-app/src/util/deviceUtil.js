// Util for use in conditional statements within code (not just css)
export const deviceSize = {
  mobile: 576,
  ipad: 768,
  ipadpro: 1025,
};
