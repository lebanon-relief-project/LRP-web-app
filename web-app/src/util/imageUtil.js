// Converts an ArrayBuffer to image data to be used in a <img />
export const getSvgImageFromArrayBuffer = (bufferData) => {
  return URL.createObjectURL(
    new Blob([new Buffer(bufferData)], { type: "image/svg+xml" })
  );
};
