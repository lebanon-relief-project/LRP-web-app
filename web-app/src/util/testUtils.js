import React from "react";
import { MemoryRouter } from "react-router-dom";
import { render } from "@testing-library/react";

export const renderWithRouter = (component, path = "/") => {
  return render(
    <MemoryRouter initialEntries={[path]}>{component}</MemoryRouter>
  );
};
