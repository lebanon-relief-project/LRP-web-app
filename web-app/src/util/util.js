export const getCardIdsFromSessionStorage = () => {
  let cardIds = [];
  let key;
  const numberOfCards = sessionStorage.length;

  for (let i = 0; i < numberOfCards; i++) {
    key = sessionStorage.key(i);

    // Should regex this to match flashcard id patterns
    if (key !== "null" && key !== "undefined") {
      cardIds.push(key);
    }
  }

  return cardIds;
};

export const getCardTitleFromSessionStorage = (id) => {
  let card = "";
  const numberOfCards = sessionStorage.length;

  for (let i = 0; i < numberOfCards; i++) {
    let key = sessionStorage.key(i);

    if (key === id) {
      card = sessionStorage.getItem(key);
    }
  }

  return card;
};

export const checkIfCardIsInSessionStorage = (id) => {
  let response = false;
  const numberOfCards = sessionStorage.length;

  for (let i = 0; i < numberOfCards; i++) {
    let key = sessionStorage.key(i);

    if (key === id) {
      response = true;
    }
  }

  return response;
};