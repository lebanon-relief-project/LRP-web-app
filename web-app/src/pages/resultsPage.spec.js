import React from "react";
import { renderWithRouter } from "../util/testUtils";
import ResultsPage from "./resultsPage";

import * as resultsService from "../services/results.service.js";
import * as util from "../util/util.js";
import { wait } from "@testing-library/react";

jest.mock("../services/results.service.js");
jest.mock("../util/util.js");

jest.mock("../components/resultsPage/ExperiencesSection", () => {
  return () => <div>Experiences Section</div>;
});

jest.mock("../components/resultsPage/ResultsSection", () => {
  return () => <div data-testid="results_section">Results Section</div>;
});

const getResultsSpy = jest.spyOn(resultsService, "getResults");
const getCardIdsFromSessionStorageSpy = jest.spyOn(
  util,
  "getCardIdsFromSessionStorage"
);

describe("the ResultsPage component", () => {
  beforeEach(() => {
    sessionStorage.clear();
    sessionStorage.setItem("key1", "value1");
    sessionStorage.setItem("key2", "value2");

    getCardIdsFromSessionStorageSpy.mockReturnValue(["key1", "key2"]);
    getResultsSpy.mockReturnValue({
      results: [
        {
          _id: "id",
          _rev: "rev",
          expl_title: "title",
          expl_body: "body",
          image: "image",
          recommendations: [
            {
              title: "some recommendation title",
              body: "some recommendation body",
            },
          ],
        },
      ],
    });
  });

  it("should match the snapshot", async () => {
    const { container } = renderWithRouter(<ResultsPage />);
    await wait();
    expect(container).toMatchSnapshot();
  });

  it("should call the results service with card id's that are in session storage", async () => {
    const { container } = renderWithRouter(<ResultsPage />);

    await wait();

    expect(getCardIdsFromSessionStorageSpy).toHaveBeenCalled();
    expect(getResultsSpy).toHaveBeenCalledWith(["key1", "key2"]);
  });

  it("should handle error response from getResults", async () => {
    getResultsSpy.mockImplementation(() => {
      throw new Error("some ugly error");
    });

    const { container, queryByTestId } = renderWithRouter(<ResultsPage />);

    await wait();

    expect(queryByTestId("results_section")).toBeFalsy();
  });
});
