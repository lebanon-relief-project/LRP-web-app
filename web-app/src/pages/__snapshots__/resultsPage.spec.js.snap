// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the ResultsPage component should match the snapshot 1`] = `
.c4 {
  margin-left: 20px;
}

.c1 {
  width: 60%;
  padding-top: 2rem;
  padding-bottom: 2rem;
  margin: 0 16.7%;
  background: inherit;
  background: #fafafa;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  border-radius: 8px;
  padding: 5%;
}

.c3 {
  font-family: Playfair Display;
  font-style: normal;
  font-weight: bold;
  font-size: 30px;
  line-height: 38px;
}

.c13 {
  position: absolute;
  bottom: 0;
  left: 50%;
  -webkit-transform: translateX(-50%);
  background-color: #FFD666;
  color: black;
  font-style: normal;
  font-weight: bold;
  padding: 0.5rem;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 90%;
  margin-bottom: 5%;
}

.c12 {
  position: absolute;
  bottom: 0;
  left: 50%;
  -webkit-transform: translateX(-50%);
  background-color: #FFD666;
  color: black;
  font-style: normal;
  font-weight: bold;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  min-width: 90%;
  max-width: 90%;
  margin-bottom: 5%;
  padding: 0.5rem;
}

.c14 {
  position: absolute;
  bottom: 0;
  left: 50%;
  -webkit-transform: translateX(-50%);
  background-color: #f0f0f0;
  color: black;
  font-style: normal;
  font-weight: bold;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  min-width: 90%;
  max-width: 90%;
  margin-bottom: 5%;
  padding: 0.5rem;
}

.c10 {
  font-family: Playfair Display;
  font-weight: bold;
  font-size: 24px;
  min-width: auto;
  line-height: 100%;
  padding: 0 0.5rem;
  text-align: center;
  margin-bottom: 0;
}

.c8 {
  position: relative;
  border: 1px solid #003A8C;
  width: 25%;
  padding: 1rem 1.2rem;
  margin: 3rem 1rem;
  height: 20%;
  font-family: Raleway;
  font-size: 16px;
  font-style: normal;
  font-weight: 400;
  line-height: 24px;
  -webkit-letter-spacing: 0em;
  -moz-letter-spacing: 0em;
  -ms-letter-spacing: 0em;
  letter-spacing: 0em;
  text-align: left;
  background-color: white;
}

.c11 {
  min-height: 7rem;
}

.c5 {
  width: 60%;
  min-width: 960px;
  padding-top: 2rem;
  padding-bottom: 2rem;
  margin: 0 16.7%;
  background: inherit;
}

.c6 {
  font-family: Playfair Display;
  font-style: normal;
  font-weight: bold;
  font-size: 30px;
  line-height: 38px;
}

.c9 {
  height: auto;
  width: 30%;
  margin: 5% 0;
}

.c9 legend {
  width: auto;
  margin: 0 auto;
}

.c7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c0 {
  max-width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

@media (max-width:1025px) {
  .c4 {
    magin-left: 0px;
  }
}

@media (max-width:576px) {
  .c4 {
    magin-left: 0px;
  }
}

@media (max-width:1025px) {
  .c1 {
    margin: 25px 18px;
    padding-top: 0px;
    padding-bottom: 0px;
    width: auto;
    max-width: 100%;
    padding: 0;
    margin: 0;
  }
}

@media (max-width:1025px) {
  .c2 {
    -webkit-flex-direction: column;
    -ms-flex-direction: column;
    flex-direction: column;
    padding: 0;
    border-radius: 0;
    margin: 0 16px;
    width: 80%;
    margin: 0 auto;
  }
}

@media (max-width:576px) {
  .c2 {
    width: 90%;
  }
}

@media (max-width:576px) {
  .c13 {
    width: 90%;
  }
}

@media (max-width:576px) {
  .c12 {
    width: 90%;
    max-width: 90%;
  }
}

@media (max-width:576px) {
  .c14 {
    width: 90%;
    max-width: 90%;
  }
}

@media (max-width:576px) {
  .c10 {
    font-size: 30px;
    line-height: 38px;
  }
}

@media (max-width:1025px) {
  .c8 {
    width: 40%;
  }
}

@media (max-width:576px) {
  .c8 {
    width: 90%;
    margin: 1rem auto;
  }
}

@media (max-width:1025px) {
  .c5 {
    margin: 25px 18px;
    padding-top: 0px;
    padding-bottom: 0px;
    width: auto;
    min-width: 0;
  }
}

@media (max-width:576px) {
  .c9 {
    width: 100%;
  }
}

@media (max-width:576px) {
  .c7 {
    -webkit-flex-direction: column;
    -ms-flex-direction: column;
    flex-direction: column;
  }
}

@media (max-width:576px) {
  .c0 {
    -webkit-flex-direction: column;
    -ms-flex-direction: column;
    flex-direction: column;
  }
}

<div>
  <div
    class="c0"
  >
    <div
      style="width: 100%;"
    >
      <div>
        Experiences Section
      </div>
      <div
        data-testid="results_section"
      >
        Results Section
      </div>
      <div
        class="c1"
      >
        <div
          class="c2"
        >
          <div>
            <h2
              class="c3"
            >
              A few more useful tips
            </h2>
          </div>
          <div
            class="c4"
          >
            <img
              alt="Person Taking Care"
              src="PersonTakingCare.svg"
            />
          </div>
        </div>
      </div>
      <div
        class="c5"
      >
        <h2
          class="c6"
        >
          Some actions you can take
        </h2>
        <div
          class="c7"
        >
          <fieldset
            class="c8 c9"
          >
            <legend
              class="c10 c9"
            >
              Support
            </legend>
            <p
              class="c11"
            >
              Surround yourself with friends and family. And remember, you can always reach outâ€¦
            </p>
            <a
              class="c12 c9"
              href="/"
            >
              Reach Out
            </a>
          </fieldset>
          <fieldset
            class="c8 c9"
          >
            <legend
              class="c10 c9"
            >
              Useful exercises
            </legend>
            <p
              class="c11"
            >
              Here are some exercises to reduce stress and regain control over your actions and thoughts.
            </p>
            <a
              class="c13"
              href="Self-Help_Exercises.pdf"
              rel="noopener noreferrer"
              target="_blank"
            >
              View exercises
            </a>
          </fieldset>
          <fieldset
            class="c8 c9"
            disabled=""
          >
            <legend
              class="c10 c9"
              disabled=""
            >
              14 days of wellbeing
            </legend>
            <p
              class="c11"
            >
              Keep your mind and body engaged by joining our 14 days of wellbeing.
            </p>
            <a
              class="c14 c9"
              disabled=""
              href="/help"
            >
              Coming Soon
            </a>
          </fieldset>
        </div>
      </div>
    </div>
  </div>
</div>
`;
